File: modules/aws/src/main/kotlin/xtdb/aws/S3.kt
@file:UseSerializers(PathWithEnvVarSerde::class, StringWithEnvVarSerde::class)

package xtdb.aws

import kotlinx.coroutines.*
import kotlinx.coroutines.future.await
import kotlinx.coroutines.future.future
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import kotlinx.serialization.UseSerializers
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider
import software.amazon.awssdk.core.async.AsyncRequestBody
import software.amazon.awssdk.core.async.AsyncResponseTransformer
import software.amazon.awssdk.services.s3.S3AsyncClient
import software.amazon.awssdk.services.s3.model.*
import xtdb.api.PathWithEnvVarSerde
import xtdb.api.StringWithEnvVarSerde
import xtdb.api.module.XtdbModule
import xtdb.api.storage.ObjectStore
import xtdb.api.storage.ObjectStore.Companion.throwMissingKey
import xtdb.api.storage.ObjectStore.StoredObject
import xtdb.api.storage.Storage.storageRoot
import xtdb.aws.s3.S3Configurator
import xtdb.multipart.IMultipartUpload
import xtdb.multipart.SupportsMultipart
import xtdb.util.asPath
import java.net.URI
import java.nio.ByteBuffer
import java.nio.file.Path
import java.util.concurrent.CompletableFuture
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Consumer
import kotlin.time.Duration.Companion.seconds

/**
 * Used to set configuration options for an S3 Object Store, which can be used as implementation of objectStore within a [xtdb.api.storage.Storage.RemoteStorageFactory].
 *
 * Requires at least a [bucket][S3.Factory.bucket] - this will need to be accessible to whichever authentication credentials you use.
 * Authentication is handled via the Default AWS Credential Provider Chain.
 * See the [AWS documentation](https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/credentials.html#credentials-default) on the various methods which you can handle authentication to be able to make use of the operations inside the modules.
 *
 * For more info on setting up the necessary infrastructure on AWS to be able to use S3 as an XTDB object store, see the section on infrastructure & setting up the AWS Cloudformation Stack within our [S3 Module Reference](https://docs.xtdb.com/config/storage/s3.html).
 *
 * Example usage, as part of a node config:
 * ```kotlin
 * Xtdb.openNode {
 *    remoteStorage(
 *       objectStore = s3(bucket = "xtdb-bucket") {
 *           prefix = Path.of("my/custom/prefix")
 *       },
 *       localDiskCache = Paths.get("test-path")
 *    ),
 *    ...
 * }
 * ```
 */
class S3(
    factory: Factory,
    private val bucket: String,
    private val prefix: Path,
) : ObjectStore, SupportsMultipart<CompletedPart> {

    private val configurator = factory.s3Configurator

    private val client =
        S3AsyncClient.builder()
            .apply {
                factory.credentials?.let { (accessKey, secretKey) ->
                    AwsBasicCredentials.create(accessKey, secretKey)
                        .let { StaticCredentialsProvider.create(it) }
                        .also { credentialsProvider(it) }
                }
                factory.endpoint?.let { endpointOverride(URI(it)) }

                configurator.configureClient(this)
            }.build()

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    override fun close() {
        runBlocking { withTimeout(5.seconds) { scope.coroutineContext.job.cancelAndJoin() } }
        client.close()
    }

    override fun startMultipart(k: Path): CompletableFuture<IMultipartUpload<CompletedPart>> = scope.future {
        val s3Key = prefix.resolve(k).toString()
        val initResp = client.createMultipartUpload {
            it.bucket(bucket)
            it.key(s3Key)
        }.await()

        val uploadId = initResp.uploadId()

        object : IMultipartUpload<CompletedPart> {
            val partNum = AtomicInteger(1)

            fun S3AsyncClient.uploadPart(body: AsyncRequestBody, configure: Consumer<UploadPartRequest.Builder>) =
                uploadPart(configure, body)

            /*
             * part-numbers have to be in ascending order, so we increment the counter synchronously.
             * caller therefore needs to call this method in order.
             */
            override fun uploadPart(buf: ByteBuffer) =
                partNum.getAndIncrement().let { partNum ->
                    scope.future {
                        val contentLength = buf.remaining().toLong()

                        val partResp = client.uploadPart(AsyncRequestBody.fromByteBuffer(buf)) {
                            it.bucket(bucket)
                            it.key(s3Key)
                            it.uploadId(uploadId)
                            it.partNumber(partNum)
                            it.contentLength(contentLength)
                        }.await()

                        CompletedPart.builder().apply {
                            partNumber(partNum)
                            eTag(partResp.eTag())
                        }.build()
                    }
                }

            override fun complete(parts: List<CompletedPart>) = scope.future {
                client.completeMultipartUpload { req ->
                    req.bucket(bucket)
                    req.key(s3Key)
                    req.uploadId(uploadId)
                    req.multipartUpload { it.parts(parts) }
                }.await()

                Unit
            }

            override fun abort() = scope.future {
                client.abortMultipartUpload {
                    it.bucket(bucket)
                    it.key(s3Key)
                    it.uploadId(uploadId)
                }.await()

                Unit
            }
        }
    }

    private fun getObjectRequest(k: Path) =
        GetObjectRequest.builder().run {
            bucket(bucket)
            key(prefix.resolve(k).toString())
            configurator.configureGet(this)
            build()
        }

    private fun <R> getObject(k: Path, responseTransformer: AsyncResponseTransformer<GetObjectResponse, R>) =
        client.getObject(getObjectRequest(k), responseTransformer)
            .exceptionally {
                if (it is NoSuchKeyException || it.cause is NoSuchKeyException) throwMissingKey(k) else throw it
            }

    override fun getObject(k: Path) = scope.future {
        getObject(k, AsyncResponseTransformer.toBytes()).await().asByteBuffer()
    }

    override fun getObject(k: Path, outPath: Path) = scope.future {
        getObject(k, AsyncResponseTransformer.toFile(outPath)).await()
        outPath
    }

    private fun S3AsyncClient.putObject(body: AsyncRequestBody, configure: Consumer<PutObjectRequest.Builder>) =
        putObject(configure, body)

    override fun putObject(k: Path, buf: ByteBuffer) =
        scope.future {
            val s3Key = prefix.resolve(k).toString()
            val headResp = runCatching {
                client.headObject {
                    it.bucket(bucket)
                    it.key(s3Key)
                    configurator.configureHead(it)
                }.await()
            }.exceptionOrNull()

            if (headResp == null) return@future
            if (headResp !is NoSuchKeyException && headResp.cause !is NoSuchKeyException) throw headResp

            val contentLength = buf.remaining().toLong()

            client.putObject(AsyncRequestBody.fromByteBuffer(buf)) {
                it.bucket(bucket)
                it.key(s3Key)
                it.contentLength(contentLength)
                configurator.configurePut(it)
            }.await()
        }

    private fun listAllObjects0(listPrefix: Path) =
        sequence {
            var continuationToken: String? = null

            while (true) {
                val listResp = runBlocking {
                    client.listObjectsV2 {
                        it.bucket(bucket)
                        it.prefix("$listPrefix/")
                        it.continuationToken(continuationToken)
                    }.await()
                }

                yieldAll(listResp.contents().map { StoredObject(prefix.relativize(it.key().asPath), it.size()) })

                if (!listResp.isTruncated) break
                continuationToken = listResp.nextContinuationToken()
            }
        }.asIterable()

    override fun listAllObjects(dir: Path) = listAllObjects0(prefix.resolve(dir))
    override fun listAllObjects() = listAllObjects0(prefix)

    // used for multipart upload testing
    fun listUploads(): Set<Path> = runBlocking {
        client.listMultipartUploads {
            it.bucket(bucket)
            it.prefix(prefix.toString())
        }.await()
            .uploads()
            .map { prefix.relativize(it.key().asPath) }
            .toSet()
    }

    override fun deleteObject(k: Path) = scope.future {
        client.deleteObject {
            it.bucket(bucket)
            it.key(prefix.resolve(k).toString())
        }

        Unit
    }


    companion object {
        @JvmStatic
        fun s3(bucket: String) = Factory(bucket)

        @Suppress("unused")
        @JvmSynthetic
        fun s3(bucket: String, configure: Factory.() -> Unit = {}) =
            s3(bucket).also(configure)

        @Serializable
        data class BasicCredentials(
            @Serializable(StringWithEnvVarSerde::class) val accessKey: String,
            @Serializable(StringWithEnvVarSerde::class) val secretKey: String
        )
    }

    @Serializable
    @SerialName("!S3")
    data class Factory(
        @Serializable(StringWithEnvVarSerde::class) val bucket: String,
        @Serializable(PathWithEnvVarSerde::class) var prefix: Path? = null,
        var credentials: BasicCredentials? = null,
        @Serializable(StringWithEnvVarSerde::class) var endpoint: String? = null,
        @Transient var s3Configurator: S3Configurator = S3Configurator.Default,
    ) : ObjectStore.Factory {

        fun prefix(prefix: Path) = apply { this.prefix = prefix }

        fun credentials(accessKey: String, secretKey: String) =
            apply { credentials = BasicCredentials(accessKey, secretKey) }

        fun endpoint(endpoint: String) = apply { this.endpoint = endpoint }

        fun s3Configurator(s3Configurator: S3Configurator) = apply { this.s3Configurator = s3Configurator }

        override fun openObjectStore() = S3(this, bucket, prefix?.resolve(storageRoot) ?: storageRoot)
    }

    /**
     * @suppress
     */
    class Registration : XtdbModule.Registration {
        override fun register(registry: XtdbModule.Registry) {
            registry.registerObjectStore(Factory::class)
        }
    }
}

File: modules/aws/src/main/clojure/xtdb/aws/s3.clj
(ns xtdb.aws.s3
  (:require [xtdb.buffer-pool :as bp]
            [xtdb.object-store :as os]
            [xtdb.util :as util])
  (:import [java.io Closeable]
           java.net.URI
           [java.nio ByteBuffer]
           [java.nio.file Path]
           [java.util ArrayList List]
           [java.util.concurrent CompletableFuture]
           [java.util.function Function]
           [software.amazon.awssdk.auth.credentials AwsBasicCredentials StaticCredentialsProvider]
           [software.amazon.awssdk.core ResponseBytes]
           [software.amazon.awssdk.core.async AsyncRequestBody AsyncResponseTransformer]
           [software.amazon.awssdk.services.s3 S3AsyncClient]
           [software.amazon.awssdk.services.s3.model AbortMultipartUploadRequest CompleteMultipartUploadRequest CompletedMultipartUpload CompletedPart CreateMultipartUploadRequest CreateMultipartUploadResponse DeleteObjectRequest GetObjectRequest HeadObjectRequest ListObjectsV2Request ListObjectsV2Response NoSuchKeyException PutObjectRequest S3Object UploadPartRequest UploadPartResponse]
           (xtdb.api.storage ObjectStore ObjectStore$StoredObject)
           [xtdb.aws S3 S3$Factory]
           [xtdb.aws.s3 S3Configurator]
           [xtdb.multipart IMultipartUpload SupportsMultipart]))

(defn- get-obj-req
  ^GetObjectRequest [{:keys [^S3Configurator configurator bucket ^Path prefix]} ^Path k]
  (let [prefixed-key (util/prefix-key prefix k)]
    (-> (GetObjectRequest/builder)
        (.bucket bucket)
        (.key (str prefixed-key))
        (doto (->> (.configureGet configurator)))
        ^GetObjectRequest (.build))))

(defn list-objects [{:keys [^S3AsyncClient client bucket ^Path prefix] :as s3-opts} continuation-token]
  )

(defn- with-exception-handler [^CompletableFuture fut ^Path k]
  (.exceptionally fut (reify Function
                        (apply [_ e]
                          (try
                            (throw (.getCause ^Exception e))
                            (catch NoSuchKeyException _
                              (throw (os/obj-missing-exception k))))))))

(defn single-object-upload
  [{:keys [^S3AsyncClient client ^S3Configurator configurator bucket ^Path prefix]} ^Path k ^ByteBuffer buf]
  (let [prefixed-key (util/prefix-key prefix k)]
    (.putObject client
                (-> (PutObjectRequest/builder)
                    (.bucket bucket)
                    (.key (str prefixed-key))
                    (doto (->> (.configurePut configurator)))
                    ^PutObjectRequest (.build))
                (AsyncRequestBody/fromByteBuffer buf))))

(defrecord MultipartUpload [^S3AsyncClient client bucket ^Path prefix ^Path k upload-id !part-number ^List !completed-parts]
  IMultipartUpload 
  (uploadPart [_  buf]
    (let [prefixed-key (util/prefix-key prefix k)
          content-length (long (.limit buf))
          part-number (int (swap! !part-number inc))]
      (-> (.uploadPart client
                       (-> (UploadPartRequest/builder)
                           (.bucket bucket)
                           (.key (str prefixed-key))
                           (.uploadId upload-id)
                           (.partNumber part-number)
                           (.contentLength content-length)
                           ^UploadPartRequest (.build))
                       (AsyncRequestBody/fromByteBuffer buf))
          (.thenApply (fn [^UploadPartResponse upload-part-response]
                        (.add !completed-parts (-> (CompletedPart/builder)
                                                   (.partNumber part-number)
                                                   (.eTag (.eTag upload-part-response))
                                                   ^CompletedPart (.build))))))))
  
  (complete [_]
    (let [prefixed-key (util/prefix-key prefix k)
          ^List !sorted-parts (sort-by (fn [^CompletedPart part] (.partNumber part)) !completed-parts)]
      (.completeMultipartUpload client
                                (-> (CompleteMultipartUploadRequest/builder)
                                    (.bucket bucket)
                                    (.key (str prefixed-key))
                                    (.uploadId upload-id)
                                    (.multipartUpload (-> (CompletedMultipartUpload/builder)
                                                          (.parts !sorted-parts)
                                                          ^CompletedMultipartUpload (.build)))
                                    ^CompleteMultipartUploadRequest (.build)))))
  
  (abort [_]
    (let [prefixed-key (util/prefix-key prefix k)]
      (.abortMultipartUpload client
                             (-> (AbortMultipartUploadRequest/builder)
                                 (.bucket bucket)
                                 (.key (str prefixed-key))
                                 (.uploadId upload-id)
                                 ^AbortMultipartUploadRequest (.build))))))

(defrecord S3ObjectStore [^S3Configurator configurator ^S3AsyncClient client bucket ^Path prefix multipart-minimum-part-size]
  ObjectStore
  (getObject [this k]
    (-> (.getObject client (get-obj-req this k) (AsyncResponseTransformer/toBytes))
        (.thenApply (reify Function
                      (apply [_ bs]
                        (.asByteBuffer ^ResponseBytes bs))))
        (with-exception-handler k)))

  (getObject [this k out-path]
    (-> (.getObject client (get-obj-req this k) out-path)
        (.thenApply (reify Function
                      (apply [_ _]
                        out-path)))
        (with-exception-handler k)))

  (putObject [this k buf]
    (let [prefixed-key (util/prefix-key prefix k)]
      (-> (.headObject client
                       (-> (HeadObjectRequest/builder)
                           (.bucket bucket)
                           (.key (str prefixed-key))
                           (doto (->> (.configureHead configurator)))
                           ^HeadObjectRequest (.build)))
          (.thenApply (fn [_resp] true))
          (.exceptionally (fn [^Exception e]
                            (let [e (.getCause e)]
                              (if (instance? NoSuchKeyException e)
                                false
                                (throw e)))))
          (.thenCompose (fn [exists?]
                          (if exists?
                            (CompletableFuture/completedFuture nil)
                            (single-object-upload this k buf)))))))

  (listAllObjects [_s]
    (letfn [(list* [cont-token]
              (lazy-seq
               (let [^ListObjectsV2Request
                     req (-> (ListObjectsV2Request/builder)
                             (.bucket bucket)
                             (.prefix (some-> prefix str))
                             (.continuationToken cont-token)
                             (.build))

                     ^ListObjectsV2Response
                     resp (.get (.listObjectsV2 ^S3AsyncClient client req))]

                 (concat (for [^S3Object object (.contents resp)]
                           (os/->StoredObject (cond->> (util/->path (.key object))
                                                prefix (.relativize prefix))
                                              (.size object)))

                         (when (.isTruncated resp)
                           (list* (.nextContinuationToken resp)))))))]
      (list* nil)))

  (deleteObject [_ k]
    (let [prefixed-key (util/prefix-key prefix k)]
      (.deleteObject client
                     (-> (DeleteObjectRequest/builder)
                         (.bucket bucket)
                         (.key (str prefixed-key))
                         ^DeleteObjectRequest (.build)))))

  SupportsMultipart
  (startMultipart [_ k]
    (let [prefixed-key (util/prefix-key prefix k)
          initiate-request (-> (CreateMultipartUploadRequest/builder)
                               (.bucket bucket)
                               (.key (str prefixed-key))
                               ^CreateMultipartUploadRequest (.build))]
      (-> (.createMultipartUpload client initiate-request)
          (.thenApply (fn [^CreateMultipartUploadResponse initiate-response]
                        (->MultipartUpload client
                                           bucket
                                           prefix
                                           k
                                           (.uploadId initiate-response)
                                           (atom 0)
                                           (ArrayList.)))))))

  Closeable
  (close [_]
    (.close client)))

(defmethod bp/->object-store-factory ::object-store [_ {:keys [bucket ^S3Configurator configurator prefix credentials endpoint]}]
  (cond-> (S3/s3 bucket)
    configurator (.s3Configurator configurator)
    prefix (.prefix (util/->path prefix))
    credentials (.credentials (:access-key credentials) (:secret-key credentials))
    endpoint (.endpoint endpoint)))

(def minimum-part-size (* 5 1024 1024))

(defn open-object-store ^ObjectStore [^S3$Factory factory]
  (let [bucket (.getBucket factory)
        configurator (.getS3Configurator factory)
        credentials (.getCredentials factory)
        endpoint (.getEndpoint factory)
        s3-client (-> (S3AsyncClient/builder)
                      (doto (cond-> credentials (doto (.credentialsProvider
                                                       (-> (AwsBasicCredentials/create (.getAccessKey credentials)
                                                                                       (.getSecretKey credentials))
                                                           (StaticCredentialsProvider/create))))
                                    endpoint (.endpointOverride (URI. endpoint))))
                      (doto (->> (.configureClient configurator)))
                      (.build))
        prefix (.getPrefix factory)
        prefix-with-version (if prefix (.resolve prefix bp/storage-root) bp/storage-root)]
  
    (->S3ObjectStore configurator
                     s3-client
                     bucket
                     prefix-with-version
                     minimum-part-size)))
